source/UF/Truncations.lagda

```diff
@@ -0,0 +1,424 @@
+      where
+       I = ap b (∥∥ₙ-rec-comp (∥∥ₙ-h-level n) (λ x → ∣ (⌜ e ⌝ x) ∣[ n ]) x)
+       II = ∥∥ₙ-rec-comp (∥∥ₙ-h-level n) (λ y → ∣ (⌜ e ⌝⁻¹ y) ∣[ n ]) (⌜ e ⌝ x)
+       III = ap (λ x → ∣ x ∣[ n ]) (inverses-are-retractions' e x)
+   G : f ∘ b ∼ id
+   G = ∥∥ₙ-ind (to-＝-of-maps-with-truncated-codomain (f ∘ b) id) G'
+    where
+     G' : (y : Y) → f (b (∣ y ∣[ n ])) ＝ (∣ y ∣[ n ])
+     G' y = f (b (∣ y ∣[ n ]))           ＝⟨ I ⟩
+            f (∣ (⌜ e ⌝⁻¹ y) ∣[ n ])     ＝⟨ II ⟩
+            (∣ ⌜ e ⌝ (⌜ e ⌝⁻¹ y) ∣[ n ]) ＝⟨ III ⟩
+            (∣ y ∣[ n ])                 ∎
+      where
+       I = ap f (∥∥ₙ-rec-comp (∥∥ₙ-h-level n) (λ y → ∣ (⌜ e ⌝⁻¹ y) ∣[ n ]) y)
+       II = ∥∥ₙ-rec-comp (∥∥ₙ-h-level n) (λ x → ∣ ⌜ e ⌝ x ∣[ n ]) (⌜ e ⌝⁻¹ y)
+       III = ap (λ y → ∣ y ∣[ n ]) (inverses-are-sections' e y)
+
+ successive-truncations-equiv : (X : 𝓤 ̇) (n : ℕ)
+                             → (∥ X ∥[ n ]) ≃ (∥ (∥ X ∥[ succ n ]) ∥[ n ])
+ successive-truncations-equiv X n = (f , (b , G) , (b , H))
+  where
+   f : ∥ X ∥[ n ] → ∥ ∥ X ∥[ succ n ] ∥[ n ]
+   f = ∥∥ₙ-rec (∥∥ₙ-h-level n) (λ x → ∣ ∣ x ∣[ succ n ] ∣[ n ])
+   b : ∥ ∥ X ∥[ succ n ] ∥[ n ] → ∥ X ∥[ n ]
+   b = ∥∥ₙ-rec (∥∥ₙ-h-level n) canonical-pred-map
+   G : f ∘ b ∼ id
+   G = ∥∥ₙ-ind (λ s → hlevels-are-closed-under-id n (∥∥ₙ-h-level n) (f (b s)) s)
+               (∥∥ₙ-ind (λ t → hlevels-are-closed-under-id n
+                                (hlevels-are-closed-under-id n
+                                (∥∥ₙ-h-level n) (f (b (∣ t ∣[ n ])))
+                                (∣ t ∣[ n ])))
+                         G')
+    where
+     G' : (x : X)
+        → f (b (∣ ∣ x ∣[ succ n ] ∣[ n ])) ＝ (∣ ∣ x ∣[ succ n ] ∣[ n ])
+     G' x = f (b (∣ ∣ x ∣[ succ n ] ∣[ n ]))         ＝⟨ I ⟩
+            f (canonical-pred-map (∣ x ∣[ succ n ])) ＝⟨ II ⟩
+            f (∣ x ∣[ n ])                           ＝⟨ III ⟩
+            (∣ ∣ x ∣[ succ n ] ∣[ n ])               ∎
+      where
+       I = ap f (∥∥ₙ-rec-comp (∥∥ₙ-h-level n) canonical-pred-map
+                              (∣ x ∣[ succ n ]))
+       II = ap f (canonical-pred-map-comp x)
+       III = ∥∥ₙ-rec-comp (∥∥ₙ-h-level n) (λ x → ∣ ∣ x ∣[ succ n ] ∣[ n ]) x
+   H : b ∘ f ∼ id
+   H = ∥∥ₙ-ind (λ s → hlevels-are-closed-under-id n (∥∥ₙ-h-level n) (b (f s)) s)
+               H'
+    where
+     H' : (x : X) → b (f (∣ x ∣[ n ])) ＝ (∣ x ∣[ n ])
+     H' x = b (f (∣ x ∣[ n ]))                   ＝⟨ I ⟩
+            b (∣ ∣ x ∣[ succ n ] ∣[ n ])         ＝⟨ II ⟩
+            canonical-pred-map (∣ x ∣[ succ n ]) ＝⟨ canonical-pred-map-comp x ⟩
+            (∣ x ∣[ n ])                         ∎
+      where
+       I = ap b (∥∥ₙ-rec-comp (∥∥ₙ-h-level n) (λ x → ∣ ∣ x ∣[ succ n ] ∣[ n ]) x)
+       II = ∥∥ₙ-rec-comp (∥∥ₙ-h-level n) canonical-pred-map (∣ x ∣[ succ n ])
+
+\end{code}
+
+We now define an equivalence that characterizes the truncated identity type
+under the assumption of univalence (see the agda unimath library for a more
```

Comment:
What I meant was for you to acknowledge/give credit because the proof you give is essentially taken from the agda-unimath library.

Reply:
Gotcha.